File: android/app/src/main/java/com/tacmap/settings/OpsecSettings.kt
package com.tacmap.settings

import com.tacmap.localization.Messages
import com.tacmap.localization.LocalizedMessage

import com.tacmap.localization.L10n

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import com.tacmap.sync.RelayEndpointPolicy
import com.tacmap.util.DurablePreferenceCommit
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Coordinate format used for the map header's primary readout. The persisted
 *  enum name is intentionally stable; unknown or absent values use MGRS so
 *  existing installs keep the app's historical default. */
enum class CoordinateDisplayType(val displayName: String) {
    MGRS("MGRS"),
    WGS84("WGS84"),
    UTM("UTM");

    companion object {
        fun fromPersisted(value: String?): CoordinateDisplayType =
            entries.firstOrNull { it.name == value } ?: MGRS
    }
}

/** Map-up behavior. Persisted names are stable; absent/unknown values preserve
 * the existing north-facing, gesture-rotatable mode. */
enum class MapOrientationMode(private val displayNameKey: String) {
    NORTH_UP("North Up"),
    HEADING_UP("Heading Up");

    val displayName: String get() = L10n.text(displayNameKey)

    companion object {
        fun fromPersisted(value: String?): MapOrientationMode =
            entries.firstOrNull { it.name == value } ?: NORTH_UP
    }
}

/** User-selected cadence for screen-off Unit Sync location updates. Persisting
 * minute values keeps the preference stable if enum names change. */
enum class BackgroundUnitSyncInterval(val minutes: Int, private val displayNameKey: String) {
    ONE_MINUTE(1, "Every minute"),
    FIVE_MINUTES(5, "Every 5 minutes"),
    FIFTEEN_MINUTES(15, "Every 15 minutes"),
    THIRTY_MINUTES(30, "Every 30 minutes"),
    SIXTY_MINUTES(60, "Every 60 minutes");

    val displayName: String get() = L10n.text(displayNameKey)

    companion object {
        val DEFAULT = FIFTEEN_MINUTES

        fun fromPersisted(minutes: Int): BackgroundUnitSyncInterval =
            entries.firstOrNull { it.minutes == minutes } ?: DEFAULT
    }
}

/**
 * App-scoped general / OPSEC / privacy settings. Backed by app-private prefs,
 * exposed as StateFlows so UI, Activity window, and networking layer
 * all see the same source of truth.
 *
 * Fresh installs favour a dark-by-default network posture:
 *  - [blockScreenCapture] ON - map (with live position) stays out of
 *    recents thumbnail and screenshots/recordings by default.
 *  - [onlineLookups] OFF - place / elevation / weather / terrain providers
 *    receive nothing until the user opts in.
 *  - [onlineBasemaps] OFF - no online raster tile is requested until the user
 *    opts in or loads an imported offline map.
 *  - [backgroundUnitSyncLocation] OFF - screen-off location sharing requires
 *    explicit user consent.
 *  Both network features remain independently switchable. Stored choices are
 *  read before these fallbacks, so an update never overwrites an existing
 *  user's explicit settings.
 *
 * The "require auth to decrypt" toggle deliberately isn't here. It has to stay
 * in lockstep with which Keystore KEK the data key is wrapped under, so it
 * lives in com.tacmap.util.DataKey and prefs would only be a second, drifting
 * copy of the truth.
 */
class OpsecSettings(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences("opsec", Context.MODE_PRIVATE)
    private val persistedNetworkPreferences = resolveNetworkPreferences(prefs.all)
    private val persistedRelayPreference = RelayEndpointPolicy.resolvePersisted(
        rawValue = prefs.getString(KEY_RELAY, null),
        defaultEndpoint = DEFAULT_RELAY,
    )

    private val _blockScreenCapture = MutableStateFlow(prefs.getBoolean(KEY_SCREEN, true))
    val blockScreenCapture: StateFlow<Boolean> = _blockScreenCapture.asStateFlow()

    private val _onlineLookups = MutableStateFlow(persistedNetworkPreferences.onlineLookups)
    val onlineLookups: StateFlow<Boolean> = _onlineLookups.asStateFlow()

    private val _onlineBasemaps = MutableStateFlow(persistedNetworkPreferences.onlineBasemaps)
    val onlineBasemaps: StateFlow<Boolean> = _onlineBasemaps.asStateFlow()

    private val _backgroundUnitSyncLocation = MutableStateFlow(
        persistedNetworkPreferences.backgroundUnitSyncLocation
    )
    val backgroundUnitSyncLocation: StateFlow<Boolean> =
        _backgroundUnitSyncLocation.asStateFlow()

    private val _backgroundUnitSyncInterval = MutableStateFlow(
        BackgroundUnitSyncInterval.fromPersisted(
            prefs.getInt(
                KEY_BACKGROUND_UNIT_SYNC_INTERVAL_MINUTES,
                BackgroundUnitSyncInterval.DEFAULT.minutes,
            )
        )
    )
    val backgroundUnitSyncInterval: StateFlow<BackgroundUnitSyncInterval> =
        _backgroundUnitSyncInterval.asStateFlow()

    private val _primaryCoordinateType = MutableStateFlow(
        CoordinateDisplayType.fromPersisted(prefs.getString(KEY_COORDINATE_TYPE, null))
    )
    val primaryCoordinateType: StateFlow<CoordinateDisplayType> =
        _primaryCoordinateType.asStateFlow()

    private val _mapOrientationMode = MutableStateFlow(
        MapOrientationMode.fromPersisted(prefs.getString(KEY_MAP_ORIENTATION, null))
    )
    val mapOrientationMode: StateFlow<MapOrientationMode> = _mapOrientationMode.asStateFlow()

    private val _relayUrl = MutableStateFlow(persistedRelayPreference.endpoint)
    val relayUrl: StateFlow<String> = _relayUrl.asStateFlow()

    private val _persistenceIssue = MutableStateFlow<LocalizedMessage?>(null)
    /** Non-null means the requested value was rejected and the last durable
     * value remains active. The UI must never imply the safer value stuck. */
    val persistenceIssue: StateFlow<LocalizedMessage?> = _persistenceIssue.asStateFlow()

    private val _relayValidationIssue = MutableStateFlow<LocalizedMessage?>(null)
    val relayValidationIssue: StateFlow<LocalizedMessage?> = _relayValidationIssue.asStateFlow()

    init {
        shared = this
        if (persistedRelayPreference.needsRepair) {
            val repaired = DurablePreferenceCommit.preferences(
                preferences = prefs,
                keys = setOf(KEY_RELAY),
                mutate = { putString(KEY_RELAY, persistedRelayPreference.endpoint) },
                publish = {},
            )
            if (!repaired) {
                _persistenceIssue.value =
                    Messages.relayRecoveryFailedMessage()
            }
        }
    }

    fun setBlockScreenCapture(value: Boolean): Boolean = persistBoolean(
        KEY_SCREEN,
        value,
        _blockScreenCapture,
    )

    fun setOnlineLookups(value: Boolean): Boolean = persistBoolean(
        KEY_ONLINE,
        value,
        _onlineLookups,
    )

    fun setOnlineBasemaps(value: Boolean): Boolean = persistBoolean(
        KEY_BASEMAPS,
        value,
        _onlineBasemaps,
    )

    fun setBackgroundUnitSyncLocation(value: Boolean): Boolean = persistBoolean(
        KEY_BACKGROUND_UNIT_SYNC_LOCATION,
        value,
        _backgroundUnitSyncLocation,
    )

    fun setBackgroundUnitSyncInterval(value: BackgroundUnitSyncInterval): Boolean =
        persist(
            key = KEY_BACKGROUND_UNIT_SYNC_INTERVAL_MINUTES,
            mutate = { putInt(KEY_BACKGROUND_UNIT_SYNC_INTERVAL_MINUTES, value.minutes) },
            publish = { _backgroundUnitSyncInterval.value = value },
        )

    fun setPrimaryCoordinateType(value: CoordinateDisplayType): Boolean =
        persist(
            key = KEY_COORDINATE_TYPE,
            mutate = { putString(KEY_COORDINATE_TYPE, value.name) },
            publish = { _primaryCoordinateType.value = value },
        )

    fun setMapOrientationMode(value: MapOrientationMode): Boolean =
        persist(
            key = KEY_MAP_ORIENTATION,
            mutate = { putString(KEY_MAP_ORIENTATION, value.name) },
            publish = { _mapOrientationMode.value = value },
        )

    fun setRelayUrl(value: String): Boolean {
        val clean = when (val result = RelayEndpointPolicy.normalize(value)) {
            is RelayEndpointPolicy.Result.Valid -> result.endpoint
            is RelayEndpointPolicy.Result.Invalid -> {
                _relayValidationIssue.value = result.pendingMessage
                return false
            }
        }
        val committed = persist(
            key = KEY_RELAY,
            mutate = { putString(KEY_RELAY, clean) },
            publish = { _relayUrl.value = clean },
        )
        if (committed) _relayValidationIssue.value = null
        return committed
    }

    fun resetRelayUrl(): Boolean = setRelayUrl(DEFAULT_RELAY)

    fun clearRelayValidationIssue() {
        _relayValidationIssue.value = null
    }

    private fun persistBoolean(
        key: String,
        value: Boolean,
        state: MutableStateFlow<Boolean>,
    ): Boolean = persist(
        key = key,
        mutate = { putBoolean(key, value) },
        publish = { state.value = value },
    )

    @SuppressLint("ApplySharedPref")
    private fun persist(
        key: String,
        mutate: SharedPreferences.Editor.() -> SharedPreferences.Editor,
        publish: () -> Unit,
    ): Boolean {
        val committed = DurablePreferenceCommit.preferences(
            preferences = prefs,
            keys = setOf(key),
            mutate = mutate,
            publish = publish,
        )
        _persistenceIssue.value = if (committed) {
            null
        } else {
            Messages.privacySettingSaveFailedMessage()
        }
        return committed
    }

    companion object {
        /** App-scoped instance, set on construction. Lets the networking layer
         *  (Weather / Elevation / Terrain services, no DI context) check the
         *  [onlineLookups] OPSEC gate before making outbound requests.
         *  Mirrors iOS `OpsecSettings.shared`. */
        @Volatile var shared: OpsecSettings? = null

        /** Default relay. Self-hosters can point this elsewhere in settings. */
        const val DEFAULT_RELAY = "wss://tacmap-sync.christianbrooker.workers.dev"
        internal const val DEFAULT_ONLINE_LOOKUPS = false
        internal const val DEFAULT_ONLINE_BASEMAPS = false
        internal const val DEFAULT_BACKGROUND_UNIT_SYNC_LOCATION = false
        private const val KEY_SCREEN = "block_screen_capture"
        private const val KEY_ONLINE = "online_lookups"
        private const val KEY_BASEMAPS = "online_basemaps"
        private const val KEY_BACKGROUND_UNIT_SYNC_LOCATION = "background_unit_sync_location"
        private const val KEY_BACKGROUND_UNIT_SYNC_INTERVAL_MINUTES =
            "background_unit_sync_interval_minutes"
        private const val KEY_COORDINATE_TYPE = "primary_coordinate_type"
        private const val KEY_MAP_ORIENTATION = "map_orientation_mode"
        private const val KEY_RELAY = "relay_url"

        internal data class NetworkPreferences(
            val onlineLookups: Boolean,
            val onlineBasemaps: Boolean,
            val backgroundUnitSyncLocation: Boolean,
        )

        /** Pure resolver used by production construction and JVM regression
         * tests. Stored booleans always win; missing or malformed values use
         * the dark-by-default fresh-install posture. */
        internal fun resolveNetworkPreferences(values: Map<String, *>): NetworkPreferences =
            NetworkPreferences(
                onlineLookups = values[KEY_ONLINE] as? Boolean ?: DEFAULT_ONLINE_LOOKUPS,
                onlineBasemaps = values[KEY_BASEMAPS] as? Boolean ?: DEFAULT_ONLINE_BASEMAPS,
                backgroundUnitSyncLocation =
                    values[KEY_BACKGROUND_UNIT_SYNC_LOCATION] as? Boolean
                        ?: DEFAULT_BACKGROUND_UNIT_SYNC_LOCATION,
            )
    }
}

File: ios/TacticalMaps/App/OpsecSettings.swift
import Foundation
import Combine

/// Coordinate representation used for the map header's primary readout.
/// Raw values are stable because they are persisted in UserDefaults.
enum CoordinateDisplayFormat: String, CaseIterable, Identifiable {
    case mgrs
    case wgs84
    case utm

    struct Resolved: Equatable {
        let format: CoordinateDisplayFormat
        let text: String
    }

    static let defaultsKey = "map.coordinateDisplayFormat"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .mgrs: "MGRS"
        case .wgs84: "WGS84"
        case .utm: "UTM"
        }
    }

    static func stored(in defaults: UserDefaults = .standard) -> Self {
        guard let rawValue = defaults.string(forKey: defaultsKey),
              let stored = Self(rawValue: rawValue) else {
            // Preserve the header users already know when upgrading.
            return .mgrs
        }
        return stored
    }

    func persist(in defaults: UserDefaults = .standard) {
        defaults.set(rawValue, forKey: Self.defaultsKey)
    }

    /// Resolves the selected format to visible text. UTM is unavailable in the
    /// polar caps; falling back to WGS84 keeps the primary readout useful there.
    func resolve(mgrs: String, wgs84: String, utm: String?) -> Resolved {
        switch self {
        case .mgrs:
            return Resolved(format: .mgrs, text: mgrs)
        case .wgs84:
            return Resolved(format: .wgs84, text: wgs84)
        case .utm:
            guard let utm else {
                return Resolved(format: .wgs84, text: wgs84)
            }
            let trimmed = utm.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty, !trimmed.hasPrefix("N/A") else {
                return Resolved(format: .wgs84, text: wgs84)
            }
            return Resolved(format: .utm, text: trimmed)
        }
    }
}

/// How the map's upward direction is controlled. Raw values are stable because
/// they are persisted in UserDefaults.
enum MapOrientationMode: String, CaseIterable, Identifiable {
    case northUp
    case headingUp

    static let defaultsKey = "map.orientationMode"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .northUp: "North Up"
        case .headingUp: "Heading Up"
        }
    }

    static func stored(in defaults: UserDefaults = .standard) -> Self {
        guard let rawValue = defaults.string(forKey: defaultsKey),
              let stored = Self(rawValue: rawValue) else {
            // Preserve the existing gesture-controlled, north-up starting state.
            return .northUp
        }
        return stored
    }

    func persist(in defaults: UserDefaults = .standard) {
        defaults.set(rawValue, forKey: Self.defaultsKey)
    }
}

/// User-selectable cadence for screen-off Unit Sync location sharing.
/// Raw minute values are stable because they are persisted in UserDefaults.
enum BackgroundUnitSyncInterval: Int, CaseIterable, Identifiable {
    case oneMinute = 1
    case fiveMinutes = 5
    case fifteenMinutes = 15
    case thirtyMinutes = 30
    case sixtyMinutes = 60

    static let defaultsKey = "opsec.backgroundUnitSyncIntervalMinutes"
    static let defaultValue: Self = .fifteenMinutes

    var id: Int { rawValue }
    var seconds: TimeInterval { TimeInterval(rawValue * 60) }

    var label: String {
        rawValue == 1 ? "Every minute" : "Every \(rawValue) minutes"
    }

    static func stored(in defaults: UserDefaults = .standard) -> Self {
        guard let stored = Self(rawValue: defaults.integer(forKey: defaultsKey)) else {
            return defaultValue
        }
        return stored
    }

    func persist(in defaults: UserDefaults = .standard) {
        defaults.set(rawValue, forKey: Self.defaultsKey)
    }
}

/// App-wide OPSEC / privacy settings. Persisted in UserDefaults and observable
/// by SwiftUI. Fresh installs favour a dark-by-default network posture:
/// privacyScreen ON - opaque cover redacts map (with live position)
/// in app-switcher snapshot whenever app isn't active.
/// onlineLookups OFF - providers receive no query or coordinate until opt-in.
/// onlineBasemaps OFF - providers receive no viewed tile until opt-in.
/// backgroundUnitSyncLocation OFF - screen-off position egress requires opt-in.
/// Every network feature remains independently switchable. Existing UserDefaults
/// values take precedence, so an update preserves each user's explicit choices.
final class OpsecSettings: ObservableObject {
    static let shared = OpsecSettings()
    static let defaultOnlineLookups = false
    static let defaultOnlineBasemaps = false
    /// Existing users must explicitly opt in before an update begins sharing
    /// their position after the app leaves the foreground.
    static let defaultBackgroundUnitSyncLocation = false

    @Published private(set) var privacyScreen: Bool
    @Published private(set) var onlineLookups: Bool
    @Published private(set) var onlineBasemaps: Bool
    @Published private(set) var backgroundUnitSyncLocation: Bool
    @Published private(set) var backgroundUnitSyncInterval: BackgroundUnitSyncInterval
    @Published private(set) var relayURL: String
    @Published private(set) var coordinateDisplayFormat: CoordinateDisplayFormat
    @Published private(set) var mapOrientationMode: MapOrientationMode
    @Published private(set) var persistenceIssue: String?
    @Published private(set) var relayValidationIssue: String?

    // Host only, no "/room/" - SyncManager appends the full "/room/<id>" path
    // itself. (Matches SyncManager.relayBase; a trailing "/room/" here would
    // double up the path and the relay would 404 the socket.)
    static let defaultRelay = "wss://tacmap-sync.christianbrooker.workers.dev"

    private let defaults: UserDefaults
    private let synchronize: (UserDefaults) -> Bool

    init(
        defaults: UserDefaults = .standard,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        synchronize: @escaping (UserDefaults) -> Bool = { $0.synchronize() }
    ) {
        self.defaults = defaults
        self.synchronize = synchronize
        persistenceIssue = nil
        relayValidationIssue = nil
        privacyScreen = defaults.object(forKey: Keys.privacy) as? Bool ?? true
        onlineLookups = defaults.object(forKey: Keys.online) as? Bool ?? Self.defaultOnlineLookups
        onlineBasemaps = defaults.object(forKey: Keys.basemaps) as? Bool ?? Self.defaultOnlineBasemaps
        backgroundUnitSyncLocation = defaults.object(forKey: Keys.backgroundUnitSyncLocation) as? Bool
            ?? Self.defaultBackgroundUnitSyncLocation
        backgroundUnitSyncInterval = BackgroundUnitSyncInterval.stored(in: defaults)
        let storedRelay = defaults.string(forKey: Keys.relay)
        let relayResolution = RelayEndpointPolicy.resolvePersisted(
            storedRelay,
            defaultEndpoint: Self.defaultRelay
        )
        relayURL = relayResolution.endpoint
        coordinateDisplayFormat = CoordinateDisplayFormat.stored(in: defaults)
        mapOrientationMode = MapOrientationMode.stored(in: defaults)
        // Marketing-screenshot mode: the store XCUITest sets this env var so the
        // shots explicitly request online tiles/lookups regardless of saved state.
        // Never set in production (env vars can't be injected into a
        // shipped app), so this is inert outside the screenshot harness.
        if environment["TACMAP_UITEST_ONLINE"] == "1" {
            onlineBasemaps = true
            onlineLookups = true
        } else if environment["TACMAP_UITEST_OFFLINE_BASEMAP"] == "1" {
            // GeoPDF slide: force the online basemap OFF (authoritatively, over any
            // value a prior run persisted) so the imported PDF sheet is what shows.
            onlineBasemaps = false
        }
        if relayResolution.needsRepair,
           !persist(relayResolution.endpoint, key: Keys.relay, verify: {
               self.defaults.string(forKey: Keys.relay) == relayResolution.endpoint
           }) {
            persistenceIssue = "The saved Unit Sync relay was unsafe or obsolete. TacMap is using its secure default for this run, but could not repair the saved setting."
        }
    }

    @discardableResult
    func setPrivacyScreen(_ value: Bool) -> Bool {
        guard persist(value, key: Keys.privacy, verify: {
            self.defaults.object(forKey: Keys.privacy) as? Bool == value
        }) else { return false }
        privacyScreen = value
        return true
    }

    @discardableResult
    func setOnlineLookups(_ value: Bool) -> Bool {
        guard persist(value, key: Keys.online, verify: {
            self.defaults.object(forKey: Keys.online) as? Bool == value
        }) else { return false }
        onlineLookups = value
        return true
    }

    @discardableResult
    func setOnlineBasemaps(_ value: Bool) -> Bool {
        guard persist(value, key: Keys.basemaps, verify: {
            self.defaults.object(forKey: Keys.basemaps) as? Bool == value
        }) else { return false }
        onlineBasemaps = value
        return true
    }

    @discardableResult
    func setBackgroundUnitSyncLocation(_ value: Bool) -> Bool {
        guard persist(value, key: Keys.backgroundUnitSyncLocation, verify: {
            self.defaults.object(forKey: Keys.backgroundUnitSyncLocation) as? Bool == value
        }) else { return false }
        backgroundUnitSyncLocation = value
        return true
    }

    @discardableResult
    func setBackgroundUnitSyncInterval(_ value: BackgroundUnitSyncInterval) -> Bool {
        guard persist(value.rawValue, key: BackgroundUnitSyncInterval.defaultsKey, verify: {
            self.defaults.integer(forKey: BackgroundUnitSyncInterval.defaultsKey) == value.rawValue
        }) else { return false }
        backgroundUnitSyncInterval = value
        return true
    }

    @discardableResult
    func setRelayURL(_ value: String) -> Bool {
        let normalized: String
        do {
            normalized = try RelayEndpointPolicy.normalize(value)
        } catch let error as RelayEndpointPolicy.ValidationError {
            relayValidationIssue = error.localizedDescription
            return false
        } catch {
            relayValidationIssue = "The relay address is not valid."
            return false
        }
        guard persist(normalized, key: Keys.relay, verify: {
            self.defaults.string(forKey: Keys.relay) == normalized
        }) else { return false }
        relayURL = normalized
        relayValidationIssue = nil
        return true
    }

    @discardableResult
    func resetRelayURL() -> Bool {
        setRelayURL(Self.defaultRelay)
    }

    func clearRelayValidationIssue() {
        relayValidationIssue = nil
    }

    @discardableResult
    func setCoordinateDisplayFormat(_ value: CoordinateDisplayFormat) -> Bool {
        guard persist(value.rawValue, key: CoordinateDisplayFormat.defaultsKey, verify: {
            self.defaults.string(forKey: CoordinateDisplayFormat.defaultsKey) == value.rawValue
        }) else { return false }
        coordinateDisplayFormat = value
        return true
    }

    @discardableResult
    func setMapOrientationMode(_ value: MapOrientationMode) -> Bool {
        guard persist(value.rawValue, key: MapOrientationMode.defaultsKey, verify: {
            self.defaults.string(forKey: MapOrientationMode.defaultsKey) == value.rawValue
        }) else { return false }
        mapOrientationMode = value
        return true
    }

    /// UserDefaults mutates its process cache before durability is known. Keep
    /// the prior exact object, require a synchronized readback, and roll back
    /// before publishing any setting to the rest of the app.
    private func persist(
        _ value: Any,
        key: String,
        verify: () -> Bool
    ) -> Bool {
        let previous = defaults.object(forKey: key)
        defaults.set(value, forKey: key)
        guard synchronize(defaults), verify() else {
            if let previous {
                defaults.set(previous, forKey: key)
            } else {
                defaults.removeObject(forKey: key)
            }
            _ = synchronize(defaults)
            persistenceIssue = "Could not save this privacy setting. The previous setting remains active; check available storage and try again."
            return false
        }
        persistenceIssue = nil
        return true
    }

    // The "require auth to decrypt" toggle deliberately isn't here. It has to
    // stay in lockstep with the access control on the Keychain item, so it lives
    // in DataKey and UserDefaults would only be a second, drifting copy.

    private enum Keys {
        static let privacy = "opsec.privacyScreen"
        static let online = "opsec.onlineLookups"
        static let basemaps = "opsec.onlineBasemaps"
        static let backgroundUnitSyncLocation = "opsec.backgroundUnitSyncLocation"
        static let relay = "opsec.relayURL"
    }
}
